File: src/sports_management/modules/notification_module/lib/data/repositories/notification_repository_impl.dart
import 'package:server_module/server_module.dart';
import '../../domain/repositories/notification_repository.dart';
import '../datasources/remote/notification_remote_data_source.dart';

class AppNotificationRepositoryImpl implements AppNotificationRepository {
  AppNotificationRepositoryImpl(this._remoteDataSource);

  final NotificationRemoteDataSource _remoteDataSource;

  @override
  Future<List<NotificationEntity>> getNotifications() async {
    final response = await _remoteDataSource.getNotifications();
    if (response.success && response.data != null) {
      try {
        final List list;
        if (response.data is List) {
          list = response.data as List;
        } else if (response.data is Map) {
          final dataMap = response.data as Map<String, dynamic>;
          list = dataMap['items'] as List? ?? [];
        } else {
          return [];
        }
        return list.map((json) {
          final model = NotificationModel.fromJson(json as Map<String, dynamic>);
          return NotificationEntity(
            id: model.id,
            userId: model.userId,
            title: model.title,
            content: model.content,
            type: model.type,
            metadata: model.metadata,
            isRead: model.isRead,
            createdAt: model.createdAt,
          );
        }).toList();
      } catch (_) {
        return [];
      }
    }
    return [];
  }

  @override
  Future<BaseResponse<dynamic>> markAsRead(String id) {
    return _remoteDataSource.markAsRead(id);
  }

  @override
  Future<BaseResponse<dynamic>> markAllAsRead() {
    return _remoteDataSource.markAllAsRead();
  }

  @override
  Future<BaseResponse<dynamic>> createNotification(Map<String, dynamic> data) {
    return _remoteDataSource.createNotification(data);
  }
}

File: src/react-staff-admin/src/features/notification/data/repositories/notification.repository_impl.ts
import { NotificationRepository } from '../../domain/repositories/notification.repository';
import { Notification } from '../../domain/entities/notification.entity';
import { NotificationRemoteDataSource, NotificationDTO } from '../datasources/notification.remote_datasource';

export class NotificationMapper {
  static toEntity(dto: NotificationDTO): Notification {
    return {
      id: dto._id || dto.id || '',
      userId: dto.userId,
      title: dto.title,
      body: dto.body,
      type: dto.type,
      isRead: dto.isRead || false,
      createdAt: dto.createdAt,
      metadata: dto.metadata
    };
  }
}

export class NotificationRepositoryImpl implements NotificationRepository {
  constructor(private remoteDataSource: NotificationRemoteDataSource) {}

  async getNotifications(): Promise<{ items: Notification[]; unreadCount: number }> {
    const { items, unreadCount } = await this.remoteDataSource.getNotifications();
    return {
      items: items.map(NotificationMapper.toEntity),
      unreadCount
    };
  }

  async markAsRead(id: string): Promise<void> {
    return this.remoteDataSource.markAsRead(id);
  }

  async markAllAsRead(): Promise<void> {
    return this.remoteDataSource.markAllAsRead();
  }

  async createNotification(payload: {
    userId?: string;
    title: string;
    body: string;
    type: Notification['type'];
    metadata?: Record<string, any>;
  }): Promise<Notification> {
    const dto = await this.remoteDataSource.createNotification(payload);
    return NotificationMapper.toEntity(dto);
  }
}
